File: 66_haspath.py
def hasPath(matrix, rows, cols, path):
	for i in range(len(matrix)):
		vis = [0] * len(matrix)
		result = onepath(matrix, rows, cols, path, i, vis)
		if result == True: return True
	return False

def onepath(matrix, rows, cols, left_path, start, vis):
	
	row = start // cols
	col = start % cols
	# print('start:', start, 'row:', row, 'col:', col)
	# print('number:', matrix[start], 'left_path:', left_path)
	# print('---------------------')
	if matrix[start] != left_path[0]: return False
	else:
		vis[start] = 1
		if len(left_path) == 1: return True
		elif len(left_path) > 1:
			up, down, left, right = False, False, False, False
			if row > 0 and vis[start - cols] == 0 : up = onepath(matrix, rows, cols, left_path[1:], start - cols, vis)
			if row < rows - 1 and vis[start + cols] == 0: down = onepath(matrix, rows, cols, left_path[1:], start + cols, vis)
			if col > 0 and vis[start - 1] == 0: left = onepath(matrix, rows, cols, left_path[1:], start - 1, vis)
			if col < cols - 1 and vis[start + 1] == 0: right = onepath(matrix, rows, cols, left_path[1:], start + 1, vis)
			result = up or down or left or right
			if result == True: return True
			else: 
				vis[start] = 0
				print("after turn left_path:", left_path)
				return False


matrix =  ['a', 'b', 'c', 'e', 's', 'f', 'c', 's', 'a,', 'd', 'e', 'e']
path = ['b','c','c','e','d']
path = ['a', 'b', 'c', 'b']
result = hasPath(matrix, 3, 4, path)
print(result)

File: 66_haspath.cpp
class Solution {
public:
    bool hasPath(char* matrix, int rows, int cols, char* str)
    {
        bool flag = false;
        for(int start = 0; start < rows * cols; ++start)
        {
            if(matrix[start] == *str)
            {
                flag = dfs(matrix, rows, cols, str, start);
                if(flag) return true;
            } 
        }
        return flag;
    }
    
    bool dfs(char *matrix, int rows, int cols, char *str, int cur)
    {
        if(*str == '\0') return true;
        if(cur < 0 || cur >= rows * cols) return false;
        if(*str != matrix[cur]) return false;
        char temp = matrix[cur];
        matrix[cur] = '*';
        bool flag = false;
        flag = dfs(matrix, rows, cols, str+1, cur - cols) || dfs(matrix, rows, cols, str + 1, cur + cols) || dfs(matrix, rows, cols, str + 1, cur - 1) || dfs(matrix, rows, cols, str + 1, cur + 1);
        if(flag == false) matrix[cur] = temp;
        return flag;
    } 
};

// 运行时间：3ms
// 占用内存：468k

// 请设计一个函数，用来判断在一个矩阵中是否存在一条包含某字符串所有字符的路径。
// 路径可以从矩阵中的任意一个格子开始，每一步可以在矩阵中向左，向右，向上，向下移动一个格子。
// 如果一条路径经过了矩阵中的某一个格子，则之后不能再次进入这个格子。 
// 例如 a b c e s f c s a d e e 这样的3 X 4 矩阵中包含一条字符串"bcced"的路径，
// 但是矩阵中不包含"abcb"路径，因为字符串的第一个字符b占据了矩阵中的第一行第二个格子之后，路径不能再次进入该格子。